ra-tls/libenclave-tls/enclave-tls.c

```diff
@@ -0,0 +1,262 @@
+typedef int (*wolfSSL_read_func)(WOLFSSL *, void *data, int);
+typedef int (*wolfSSL_write_func)(WOLFSSL *, const void *, int);
+typedef WOLFSSL_CTX *(*wolfSSL_CTX_new_func)(WOLFSSL_METHOD *);
+typedef WOLFSSL_METHOD *(*wolfTLSv1_2_client_method_func)(void);
+typedef WOLFSSL_METHOD *(*wolfTLSv1_2_server_method_func)(void);
+typedef const unsigned char *(*wolfSSL_X509_get_der_func)(WOLFSSL_X509 *,
+							  int *);
+typedef WOLFSSL_X509 *(*wolfSSL_get_peer_certificate_func)(WOLFSSL *);
+typedef int (*wolfSSL_connect_func)(WOLFSSL *);
+typedef int (*wolfSSL_negotiate_func)(WOLFSSL *);
+typedef int (*wolfSSL_set_fd_func)(WOLFSSL *, int);
+typedef WOLFSSL *(*wolfSSL_new_func)(WOLFSSL_CTX *);
+typedef void (*wolfSSL_CTX_free_func)(WOLFSSL_CTX *);
+typedef int (*wolfSSL_Debugging_ON_func)(void);
+typedef void (*wolfSSL_Debugging_OFF_func)(void);
+typedef int (*wolfSSL_Init_func)(quote_type_t);
+typedef void (*create_key_and_x509_func)(WOLFSSL_CTX *);
+typedef int (*wolfSSL_Cleanup_func)(void);
+typedef void (*printf_identity_func)(unsigned char *, int);
+typedef int (*verify_cert_extensions_func)(uint8_t *, uint32_t);
+
+// dynamically resolved functions from libra-tls-sgx libraries
+typedef struct enclave_tls_function_table {
+	printf_identity_func printf_identity;
+	wolfSSL_free_func wolfSSL_free;
+	wolfSSL_CTX_set_verify_func wolfSSL_CTX_set_verify;
+	create_key_and_x509_func create_key_and_x509;
+	wolfSSL_read_func wolfSSL_read;
+	wolfSSL_write_func wolfSSL_write;
+	wolfSSL_CTX_new_func wolfSSL_CTX_new;
+	wolfTLSv1_2_client_method_func wolfTLSv1_2_client_method;
+	wolfSSL_X509_get_der_func wolfSSL_X509_get_der;
+	wolfSSL_get_peer_certificate_func wolfSSL_get_peer_certificate;
+	wolfSSL_connect_func wolfSSL_connect;
+	wolfSSL_negotiate_func wolfSSL_negotiate;
+	wolfSSL_set_fd_func wolfSSL_set_fd;
+	wolfSSL_new_func wolfSSL_new;
+	wolfTLSv1_2_server_method_func wolfTLSv1_2_server_method;
+	wolfSSL_CTX_free_func wolfSSL_CTX_free;
+	wolfSSL_Cleanup_func wolfSSL_Cleanup;
+	wolfSSL_Debugging_ON_func wolfSSL_Debugging_ON;
+	wolfSSL_Debugging_OFF_func wolfSSL_Debugging_OFF;
+	wolfSSL_Init_func wolfSSL_Init;
+} enclave_tls_function_table;
+
+/* global function table */
+enclave_tls_function_table ftab;
+
+int cert_verify_callback(int preverify, WOLFSSL_X509_STORE_CTX *store)
+{
+	(void) preverify;
+	int ret = 0;
+	if (is_epid_ratls_cert(store->certs->buffer, store->certs->length)) {
+		void *handle = dlopen(LIBENCLAVE_TLS_SGX, RTLD_LAZY);
+		if (NULL == handle) {
+			printf("open error:%s\n", dlerror());
+			return -1;
+		}
+		verify_cert_extensions_func verify_cert_extensions =
+			dlsym(handle, "verify_sgx_cert_extensions");
+		ret = verify_cert_extensions(store->certs->buffer,
+					     store->certs->length);
+		dlclose(handle);
+	}
+
+	fprintf(stderr, "Verifying certificate extensions ... %s\n",
+		ret == 0 ? "Success" : "Failure");
+	return !ret;
+}
+
+void *enclave_tls_init(struct enclave_tls_ctx *ctx)
```

Comment:
Sense what this function is used for:

```C
#define BUILD_BUG(cond, msg) __compiletime_assert(cond, msg)

#define __compiletime_assert(condition, msg)           \
        do {                                                            \
                extern void __compiletime_assert__(void) __attribute__((__error__(msg))); \
                if (condition)                                       \
                        __compiletime_assert__();                             \
        } while (0)

void main(void)
{
        BUILD_BUG(1, "bug on!");
}
```

and use it in this way:
```C
#include <for_app_header/enclave_tls.h>

ETLS_CTX *enclave_tls_init()
{
    struct enclave_tls_ctx *ctx = calloc(1, sizeof(*ctx));
    if (!ctx)
      ...

    BUILD_BUG(sizeof(struct enclave_tls_ctx) >= ETLS_CTX, "sizeof(ETLS_CTX) is too small");
...
    return ctx;
}
```

where <for_app_header/enclave_tls.h>:
```C
typedef struct {
    uint8_t bytes[128];
} ETLS_CTX;
```